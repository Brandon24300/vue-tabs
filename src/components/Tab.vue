<template>
    <div class="vue-tabs__item">
        <slot name="title" />
        <slot />
    </div>
</template>

<script>
export default {
  name: 'vue-tabs-item',
  props: {
    title: {
      type: String,
      required: true,
    },
    isDisabled: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<style>

</style>
