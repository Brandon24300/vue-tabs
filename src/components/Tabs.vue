
<script>
export default {
  name: 'vue-tabs-container',

  props: {
    defaultIndex: {
      type: Number,
      default: 0,
    },
    transitionName: {
      type: String,
      default: '',
    },
  },

  data() {
    return {
      activeIndex: this.defaultIndex,
    };
  },

  computed: {
    tabs() {
      return this.$slots.default
        .filter(({ tag }) => tag === 'vue-component-2-vue-tabs-item');
    },
    tabNav() {
      return this.tabs
        .map(({ componentOptions }, index) => {
          const { title } = componentOptions.propsData;

          return (
            <button
              class={ this.getTabClasses(index) }
              attrs={ this.getTabAria(index) }
              onClick={ e => this.switchTab(e, index) }
            >
              { title }
            </button>
          );
        });
    },
    tabList() {
      return (
        <div
          class="vue-tabs__nav"
          role="tablist"
        >
          { this.tabNav }
        </div>
      );
    },
    tabPanels() {
      return this.tabs
        .map((panel, index) => (
          <div
            class="vue-tabs__panel"
            key={index}
            attrs={ this.getPanelAria(index) }
          >
            { panel }
          </div>
        ));
    },
  },

  methods: {
    switchTab(e, index) {
      this.activeIndex = index;
    },
    isTabActive(index) {
      return this.activeIndex === index;
    },
    getTabClasses(index) {
      return [
        'vue-tabs__nav-item',
        { '--active': this.isTabActive(index) },
      ];
    },
    getTabAria(index) {
      return {
        role: 'tab',
        tabindex: !this.isTabActive(index) && -1,
        'aria-selected': this.isTabActive(index) ? 'true' : 'false',
        'aria-controls': `tab-${index}`,
      };
    },
    getPanelAria(index) {
      return {
        role: 'tabpanel',
        tabindex: 0,
        hidden: this.activeIndex !== index,
        'aria-labelledby': `tab-${index}`,
      };
    },
  },

  render() {
    return (
      <div class="vue-tabs">
        { this.tabList }
        <transition name={ this.transitionName } mode="out-in">
          <div key={ this.activeIndex }>
            { this.tabPanels }
          </div>
        </transition>
      </div>
    );
  },
};
</script>

<style lang="scss">
.vue-tabs {
  $bl: '.vue-tabs';

  #{$bl}__nav {
    display: flex;
    align-items: baseline;
    justify-content: center;
  }
  #{$bl}__nav-item {
  }
  #{$bl}__panel {
    &[hidden="hidden"] {
      display: none;
    }
  }
  #{$bl}__panel-container {
  }

  .fade-enter-active,
  .fade-leave-active {
    transition: opacity .4s;
  }
  .fade-enter,
  .fade-leave-to {
    opacity: 0;
  }
}
</style>
