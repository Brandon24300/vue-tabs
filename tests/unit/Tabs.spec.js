
import { createLocalVue, mount, shallowMount } from '@vue/test-utils';

import Tab from '@/components/Tab.vue';
import Tabs from '@/components/Tabs.vue';

const localVue = createLocalVue();

localVue.component('tab', Tab);

describe('Tab.vue', () => {
  it('does not render anything when slot is not passed', () => {
    const wrapper = shallowMount(Tabs);

    expect(wrapper.html()).toBe(undefined);
  });

  describe('Rendering', () => {
    const wrapper = mount(Tabs, {
      slots: {
        default: `
            <tab title="tab-1"><div>tab-1-content</div></tab>
            <tab title="tab-2"><div>tab-2-content</div></tab>
            <tab title="tab-3"><div>tab-3-content</div></tab>
            <tab title="tab-4"><div>tab-4-content</div></tab>
            <div title="tab-5"><div>tab-5-content</div></div>
        `,
      },
      localVue,
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('doesn\'t render non-tab components', () => {
      expect(wrapper.vm.getTabs().length).toBe(4);
    });

    it('returns right tabs array length', () => {
      expect(wrapper.vm.getTabsLength()).toBe(4);
    });

    describe('Tabs navigation clicks', () => {
      const tabItems = wrapper.findAll('.vue-tabs__nav-item');

      for (let i = 0; i < tabItems.length; i += 1) {
        it('Switches tabs correctly', () => {
          wrapper.findAll('.vue-tabs__nav-item').at(i).trigger('click');

          expect(wrapper.vm.activeIndex).toBe(i);
          expect(tabItems.at(i).classes()).toContain('--active');
          expect(wrapper.vm.getPanelAria(i).hidden).toBe(false);
        });
      }
    });
  });
});
