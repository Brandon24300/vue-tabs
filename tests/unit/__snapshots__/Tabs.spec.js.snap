// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab.vue Rendering renders correctly 1`] = `
<div class="vue-tabs">
  <div role="tablist" aria-label="" class="vue-tabs__nav"><button role="tab" aria-selected="true" aria-controls="tab-0" class="vue-tabs__nav-item --active">tab-1</button><button role="tab" tabindex="-1" aria-selected="false" aria-controls="tab-1" class="vue-tabs__nav-item">tab-2</button><button role="tab" tabindex="-1" aria-selected="false" aria-controls="tab-2" class="vue-tabs__nav-item">tab-3</button><button role="tab" tabindex="-1" aria-selected="false" aria-controls="tab-3" class="vue-tabs__nav-item">tab-4</button></div>
  <div name="" mode="out-in">
    <div role="tabpanel" tabindex="0" aria-labelledby="tab-0" class="vue-tabs__panel">
      <div class="vue-tabs__panel-container">
        <div>tab-1-content</div>
      </div>
    </div>
    <div role="tabpanel" tabindex="0" hidden="hidden" aria-labelledby="tab-1" class="vue-tabs__panel">
      <div class="vue-tabs__panel-container">
        <div>tab-2-content</div>
      </div>
    </div>
    <div role="tabpanel" tabindex="0" hidden="hidden" aria-labelledby="tab-2" class="vue-tabs__panel">
      <div class="vue-tabs__panel-container">
        <div>tab-3-content</div>
      </div>
    </div>
    <div role="tabpanel" tabindex="0" hidden="hidden" aria-labelledby="tab-3" class="vue-tabs__panel">
      <div class="vue-tabs__panel-container">
        <div>tab-4-content</div>
      </div>
    </div>
  </div>
</div>
`;
